File: packages/cli/src/sessions/opencode.ts
/**
 * OpenCode session parser
 * Reads from ~/.local/share/opencode/storage/message/
 */

import * as fs from "node:fs";
import * as path from "node:path";
import * as os from "node:os";
import { createUnifiedMessage, type UnifiedMessage, type TokenBreakdown } from "./types.js";

interface OpenCodeMessageFile {
  id: string;
  sessionID: string;
  role: "assistant" | "user";
  modelID?: string;
  providerID?: string;
  cost: number;
  tokens?: {
    input: number;
    output: number;
    reasoning: number;
    cache: {
      read: number;
      write: number;
    };
  };
  time: {
    created: number;
    completed?: number;
  };
  agent?: string;
  mode?: string;
}

export function normalizeAgentName(agent: string): string {
  const agentLower = agent.toLowerCase();

  if (agentLower.includes("plan")) {
    if (agentLower.includes("omo") || agentLower.includes("sisyphus")) {
      return "Planner-Sisyphus";
    }
    return agent;
  }

  if (agentLower === "omo" || agentLower === "sisyphus") {
    return "Sisyphus";
  }

  return agent;
}

export function getOpenCodeStoragePath(): string {
  const dataHome = process.env.XDG_DATA_HOME || path.join(os.homedir(), ".local", "share");
  return path.join(dataHome, "opencode", "storage");
}

export function parseOpenCodeMessages(): UnifiedMessage[] {
  const storagePath = getOpenCodeStoragePath();
  const messagePath = path.join(storagePath, "message");

  if (!fs.existsSync(messagePath)) {
    return [];
  }

  const messages: UnifiedMessage[] = [];
  
  try {
    const sessionDirs = fs
      .readdirSync(messagePath)
      .map((dir) => path.join(messagePath, dir))
      .filter((dir) => fs.statSync(dir).isDirectory());

    for (const sessionDir of sessionDirs) {
      const sessionId = path.basename(sessionDir);
      const files = fs.readdirSync(sessionDir).filter((f) => f.endsWith(".json"));

      for (const file of files) {
        try {
          const content = fs.readFileSync(path.join(sessionDir, file), "utf-8");
          const msg = JSON.parse(content) as OpenCodeMessageFile;

          if (msg.role === "assistant" && msg.tokens && msg.modelID) {
            const tokens: TokenBreakdown = {
              input: msg.tokens.input || 0,
              output: msg.tokens.output || 0,
              cacheRead: msg.tokens.cache?.read || 0,
              cacheWrite: msg.tokens.cache?.write || 0,
              reasoning: msg.tokens.reasoning || 0,
            };

            const agentOrMode = msg.mode || msg.agent;
            const agent = agentOrMode ? normalizeAgentName(agentOrMode) : undefined;

            messages.push(
              createUnifiedMessage(
                "opencode",
                msg.modelID,
                msg.providerID || "unknown",
                sessionId,
                msg.time.created,
                tokens,
                msg.cost || 0,
                agent
              )
            );
          }
        } catch {
          // Skip malformed files
        }
      }
    }
  } catch {
    // Skip inaccessible directories
  }

  return messages;
}

File: packages/core/src/sessions/opencode.rs
//! OpenCode session parser
//!
//! Parses individual JSON files from ~/.local/share/opencode/storage/message/

use super::{normalize_agent_name, UnifiedMessage};
use crate::TokenBreakdown;
use serde::Deserialize;
use std::path::Path;

/// OpenCode message structure (from JSON files)
#[derive(Debug, Deserialize)]
#[allow(dead_code)]
pub struct OpenCodeMessage {
    pub id: String,
    #[serde(rename = "sessionID")]
    pub session_id: String,
    pub role: String,
    #[serde(rename = "modelID")]
    pub model_id: Option<String>,
    #[serde(rename = "providerID")]
    pub provider_id: Option<String>,
    pub cost: Option<f64>,
    pub tokens: Option<OpenCodeTokens>,
    pub time: OpenCodeTime,
    pub agent: Option<String>,
    pub mode: Option<String>,
}

#[derive(Debug, Deserialize)]
pub struct OpenCodeTokens {
    pub input: i64,
    pub output: i64,
    pub reasoning: Option<i64>,
    pub cache: OpenCodeCache,
}

#[derive(Debug, Deserialize)]
pub struct OpenCodeCache {
    pub read: i64,
    pub write: i64,
}

#[derive(Debug, Deserialize)]
#[allow(dead_code)]
pub struct OpenCodeTime {
    pub created: f64, // Unix timestamp in milliseconds (as float)
    pub completed: Option<f64>,
}

pub fn parse_opencode_file(path: &Path) -> Option<UnifiedMessage> {
    let data = std::fs::read(path).ok()?;
    let mut bytes = data;

    let msg: OpenCodeMessage = simd_json::from_slice(&mut bytes).ok()?;

    if msg.role != "assistant" {
        return None;
    }

    let tokens = msg.tokens?;
    let model_id = msg.model_id?;
    let agent_or_mode = msg.mode.or(msg.agent);
    let agent = agent_or_mode.map(|a| normalize_agent_name(&a));

    Some(UnifiedMessage::new_with_agent(
        "opencode",
        model_id,
        msg.provider_id.unwrap_or_else(|| "unknown".to_string()),
        msg.session_id.clone(),
        msg.time.created as i64,
        TokenBreakdown {
            input: tokens.input,
            output: tokens.output,
            cache_read: tokens.cache.read,
            cache_write: tokens.cache.write,
            reasoning: tokens.reasoning.unwrap_or(0),
        },
        msg.cost.unwrap_or(0.0),
        agent,
    ))
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_parse_opencode_structure() {
        let json = r#"{
            "id": "msg_123",
            "sessionID": "ses_456",
            "role": "assistant",
            "modelID": "claude-sonnet-4",
            "providerID": "anthropic",
            "cost": 0.05,
            "tokens": {
                "input": 1000,
                "output": 500,
                "reasoning": 100,
                "cache": { "read": 200, "write": 50 }
            },
            "time": { "created": 1700000000000.0 }
        }"#;

        let mut bytes = json.as_bytes().to_vec();
        let msg: OpenCodeMessage = simd_json::from_slice(&mut bytes).unwrap();

        assert_eq!(msg.model_id, Some("claude-sonnet-4".to_string()));
        assert_eq!(msg.tokens.unwrap().input, 1000);
        assert_eq!(msg.agent, None);
    }

    #[test]
    fn test_parse_opencode_with_agent() {
        let json = r#"{
            "id": "msg_123",
            "sessionID": "ses_456",
            "role": "assistant",
            "modelID": "claude-sonnet-4",
            "providerID": "anthropic",
            "agent": "OmO",
            "cost": 0.05,
            "tokens": {
                "input": 1000,
                "output": 500,
                "reasoning": 100,
                "cache": { "read": 200, "write": 50 }
            },
            "time": { "created": 1700000000000.0 }
        }"#;

        let mut bytes = json.as_bytes().to_vec();
        let msg: OpenCodeMessage = simd_json::from_slice(&mut bytes).unwrap();

        assert_eq!(msg.agent, Some("OmO".to_string()));
    }
}
